File: src/main/java/com/extendedae_plus/integration/recipeViewer/emi/HelperBoMRecipes.java
package com.extendedae_plus.integration.recipeViewer.emi;

import dev.emi.emi.api.recipe.EmiRecipe;
import dev.emi.emi.api.recipe.EmiResolutionRecipe;
import dev.emi.emi.api.stack.EmiIngredient;
import dev.emi.emi.api.stack.EmiStack;
import dev.emi.emi.bom.FoldState;
import dev.emi.emi.bom.MaterialNode;
import net.minecraft.resources.ResourceLocation;

import java.util.ArrayList;
import java.util.HashMap;

public class HelperBoMRecipes {
    public static HashMap<ResourceLocation, EmiStack> mapChildren(MaterialNode nodeParent, long batches) {
        var recipe = nodeParent.recipe;
        if (recipe == null) return new HashMap<>();
        if (nodeParent.state != FoldState.EXPANDED) return new HashMap<>();
        if (nodeParent.children == null) return new HashMap<>();

        var resultMapped = new HashMap<ResourceLocation, EmiStack>();

        nodeParent.children.forEach(child -> {
            var stack = batchAmount(child, batches);
            recipe.getInputs().forEach(input -> {
                if (!input.getEmiStacks().contains(stack)) return;
                resultMapped.put(input.getEmiStacks().getFirst().getId(), stack);
            });
        });

        return resultMapped;
    }

    public static void applyMappings(EmiRecipeAdaptable adaptable,
                                     EmiRecipe original,
                                     HashMap<ResourceLocation, EmiStack> mapper) {
        var mapped = new ArrayList<EmiIngredient>();
        original.getInputs().forEach(input -> {
            var key = input.getEmiStacks().getFirst().getId();
            var stack = mapper.get(key);
            mapped.add(stack == null ? input : stack);
        });
        adaptable.setInputs(mapped);
    }

    public static EmiStack batchAmount(MaterialNode node, long batches) {
        EmiStack original;
        if (node.recipe instanceof EmiResolutionRecipe recipeResolution)
            original = recipeResolution.stack;
        else original = node.ingredient.getEmiStacks().getFirst();
        return original.copy().setAmount(node.amount * batches);
    }
}

File: src/main/kotlin/com/extendedae_plus/integration/impl/recipeViewer/emi/HelperBoMRecipes.kt
package com.extendedae_plus.integration.impl.recipeViewer.emi

import dev.emi.emi.api.recipe.EmiRecipe
import dev.emi.emi.api.recipe.EmiResolutionRecipe
import dev.emi.emi.api.stack.EmiIngredient
import dev.emi.emi.api.stack.EmiStack
import dev.emi.emi.bom.FoldState
import dev.emi.emi.bom.MaterialNode
import net.minecraft.resources.ResourceLocation

object HelperBoMRecipes {
    fun mapChildren(nodeParent: MaterialNode, batches: Long): HashMap<ResourceLocation, EmiStack> {
        val recipe = nodeParent.recipe ?: return HashMap()
        if (nodeParent.state != FoldState.EXPANDED) return HashMap()
        val children = nodeParent.children ?: return HashMap()

        val resultMapped = HashMap<ResourceLocation, EmiStack>()

        children.forEach { child ->
            val stack = batchAmount(child, batches)
            recipe.inputs.forEach { input ->
                if (!input.emiStacks.contains(stack)) return@forEach
                resultMapped[input.emiStacks[0].id] = stack
            }
        }

        return resultMapped
    }

    fun applyMappings(
        adaptable: EmiRecipeAdaptable,
        original: EmiRecipe,
        mapper: HashMap<ResourceLocation, EmiStack>
    ) {
        val mapped = ArrayList<EmiIngredient>()
        original.inputs.forEach { input ->
            val key = input.emiStacks[0].id
            val stack = mapper[key]
            mapped.add(stack ?: input)
        }
        adaptable.inputs = mapped
    }

    fun batchAmount(node: MaterialNode, batches: Long): EmiStack {
        val recipe = node.recipe
        val original = if (recipe is EmiResolutionRecipe) recipe.stack
        else node.ingredient.emiStacks[0]
        return original.copy().setAmount(node.amount * batches)
    }
}
